### ISSUE TITLE
Question: Support for multiple input addresses?

### ISSUE THREAD

**Issue by cybermcm** [pitkley/dfw#89]
Hi,

Love your project, I use it for my personal server, works great.
Question: Is it possible to restrict input to multiple IPs, like 
> -A INPUT -p tcp --dport 25 -m conntrack --ctstate NEW,ESTABLISHED -s 127.0.0.0/8,199.55.55.55,172.16.0.0/12 -j ACCEPT"

If I put this on top (initialization section) then it's overruled by the www->docker section if I understand iptables correct

**Comment by pitkley** [comment]
Hi @cybermcm, great to hear that you find use in dfw!

When you say "www->docker section", what exactly are you referring to by "www->docker"? Everything you put into the initialization section will be passed on to iptables as-is, so if the rule is valid, it should just work.

One potential issue could be the way you run the Docker daemon: I just realized that the README does not inform users that they should run Docker without the iptables-handling integrated in the daemon, by launching `dockerd` with the `--iptables=false` flag set (see also the [official documentation](https://docs.docker.com/network/iptables/#prevent-docker-from-manipulating-iptables)).

It might also be helpful if you can post your iptables contents, should the previous points not resolve your issue:

```
iptables -nvL
iptables -t nat -nvL
```

(Be sure to strip the output of any sensitive data, like personal IPs you might not want to be documented here.)

**Comment by cybermcm** [comment]
Maybe a misunderstanding because of my lack of iptables know-how.
The iptables output (shortened)
```
root@srv01:/opt# iptables -nvL
Chain INPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
  233 81913 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  lo     *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     icmp --  *      *       0.0.0.0/0            0.0.0.0/0            icmptype 8
    1    60 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:22 ctstate NEW,ESTABLISHED
    0     0 ACCEPT     tcp  --  *      *       127.0.0.0/8          0.0.0.0/0            tcp dpt:25 ctstate NEW,ESTABLISHED
    0     0 ACCEPT     tcp  --  *      *       111.11.11.11         0.0.0.0/0            tcp dpt:25 ctstate NEW,ESTABLISHED
    0     0 ACCEPT     tcp  --  *      *       172.16.0.0/12        0.0.0.0/0            tcp dpt:25 ctstate NEW,ESTABLISHED
    7   396 DFWRS_INPUT  all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
  124 12965 DFWRS_FORWARD  all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain OUTPUT (policy ACCEPT 246 packets, 92068 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all  --  *      lo      0.0.0.0/0            0.0.0.0/0

Chain DFWRS_FORWARD (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            state INVALID
  109 11981 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  br-2ae1c2fd8727 ens3    0.0.0.0/0            0.0.0.0/0
    4   266 ACCEPT     all  --  br-15b0c160fb5a ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-480a3d62c58f ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-182544ce66dd ens3    0.0.0.0/0            0.0.0.0/0
   10   666 ACCEPT     all  --  br-b08c9ac12933 ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-e4e931f0d1f6 ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-683fab2c3c96 ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.15          tcp dpt:53
    0     0 ACCEPT     udp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.15          udp dpt:53
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.1           tcp dpt:25
    1    52 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.1           tcp dpt:587
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.1           tcp dpt:993
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.1           tcp dpt:995
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.0.6           tcp dpt:19132
    0     0 ACCEPT     udp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.0.6           udp dpt:19132
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.2           tcp dpt:1194
    0     0 ACCEPT     udp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.2           udp dpt:1194
    0     0 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.10          tcp dpt:50455
    0     0 ACCEPT     udp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.10          udp dpt:50455
    0     0 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.7           tcp dpt:50456
    0     0 ACCEPT     udp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.7           udp dpt:50456
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.0.4           tcp dpt:445
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.0.8           tcp dpt:3128
    0     0 ACCEPT     udp  --  ens3   br-2ae1c2fd8727  0.0.0.0/0            172.21.0.2           udp dpt:9987
    0     0 ACCEPT     tcp  --  ens3   br-2ae1c2fd8727  0.0.0.0/0            172.21.0.2           tcp dpt:10011
    0     0 ACCEPT     tcp  --  ens3   br-2ae1c2fd8727  0.0.0.0/0            172.21.0.2           tcp dpt:30033
    0     0 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.6           tcp dpt:80
    0     0 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.6           tcp dpt:443
    0     0 ACCEPT     all  --  docker0 ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain DFWRS_INPUT (1 references)
 pkts bytes target     prot opt in     out     source               destination
    3   156 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            state INVALID
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  br-2ae1c2fd8727 *       0.0.0.0/0            0.0.0.0/0
    1    60 ACCEPT     all  --  br-15b0c160fb5a *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-480a3d62c58f *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-182544ce66dd *       0.0.0.0/0            0.0.0.0/0
    3   180 ACCEPT     all  --  br-b08c9ac12933 *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-e4e931f0d1f6 *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-683fab2c3c96 *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  docker0 *       0.0.0.0/0            0.0.0.0/0
```
My main goal is to secure my mail server. The mail server should only be reachable on port 25 by my MX (e.g. 111.11.11.11 and docker container). This is the input rule.
Than there is the DFWRS_FORWARD rule:
> 0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.1           tcp dpt:25

This rule has no IP restrictions, so I think any traffic for port 25 will be forwarded to the docker container.
But after consultation of Wikipedia:
>Each rule in a chain contains the specification of which packets it matches. It may also contain a target (used for extensions) or verdict (one of the built-in decisions). As a packet traverses a chain, each rule in turn is examined. If a rule does not match the packet, the packet is passed to the next rule. If a rule does match the packet, the rule takes the action indicated by the target/verdict, which may result in the packet being allowed to continue along the chain or may not.

it seems that I already accomplished my "mission".

**Comment by pitkley** [comment]
> This rule has no IP restrictions, so I think any traffic for port 25 will be forwarded to the docker container.

This observation is correct, but your follow-up conclusion might mislead you. It is in fact true that when an iptables rule matches, and the action is terminal (like `ACCEPT`, `DROP` and `REJECT`), that it will be the last rule evaluated by iptables. What you need to care about is when the rule _doesn't_ match. In this case iptables will continue to match rules, falling back to the default policy of the chain (which is shown at the start of the output: `Chain INPUT (policy DROP 0 packets, 0 bytes)`).

If I'm not mistaken, judging by your iptables-output, your exposed port 25 is world-reachable because there is no rule forbidding it, but there is a rule matching it in the `DFWRS_FORWARD` chain -- and from what I understood you don't want the port to be publicly reachable. Before continuing on I'd suggest you verify whether the port is open to the world or not -- the most simple way is probably to search for an online port checker and entering your IP and port 25.

Without seeing your configuration, I assume that you have created a `wider_world_to_container` rule for your SMTP-gateway since you need it to be externally reachable from a specific IP, correct? If so I assume that the features provided by dfw do not cover your use-case directly: when exposing a port to the wider world, it currently is always exposed to the entire IP-space and can't be directly restricted.

What you can do though is take your initialization rule and flip it around: instead of explicitly allowing access from `111.11.11.11` to your SMTP-port (which is covered by the dfw-rules already), forbid access to port 25 to everything that _isn't_ `111.11.11.11`:

```
-A INPUT -p tcp --dport 25 ! -s 111.11.11.11/32 -j DROP
```

Let me know if that helps your issue. If that covers your use-case, I'll create a follow-up ticket (or relabel this one) to track implementing a feature that allows you to specify the IP-range in the dfw-rules directly!

**Comment by cybermcm** [comment]
you perfectly summarized my use case and your are right, my port 25 is still open to the public.
I tried your suggestion with the invert rule but according to 2 different check sites my port 25 is still open to the public
```
Chain INPUT (policy DROP 20 packets, 1784 bytes)
 pkts bytes target     prot opt in     out     source               destination
14718 5013K ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  lo     *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     icmp --  *      *       0.0.0.0/0            0.0.0.0/0            icmptype 8
   13   764 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:22 ctstate NEW,ESTABLISHED
   21  1120 DROP       tcp  --  *      *      !127.0.0.0/8          0.0.0.0/0            tcp dpt:25
    0     0 DROP       tcp  --  *      *      !111.11.11.11         0.0.0.0/0            tcp dpt:25
 1789  108K DFWRS_INPUT  all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
 4398 1082K DFWRS_FORWARD  all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain OUTPUT (policy ACCEPT 16177 packets, 5105K bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all  --  *      lo      0.0.0.0/0            0.0.0.0/0

Chain DFWRS_FORWARD (1 references)
 pkts bytes target     prot opt in     out     source               destination
   12   480 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            state INVALID
 3405 1020K ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED
  665 39727 ACCEPT     all  --  br-b08c9ac12933 ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-480a3d62c58f ens3    0.0.0.0/0            0.0.0.0/0
   19  1070 ACCEPT     all  --  br-182544ce66dd ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-e4e931f0d1f6 ens3    0.0.0.0/0            0.0.0.0/0
  263 19184 ACCEPT     all  --  br-15b0c160fb5a ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-2ae1c2fd8727 ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-683fab2c3c96 ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.15          tcp dpt:53
   10   755 ACCEPT     udp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.15          udp dpt:53
   13   780 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.1           tcp dpt:25
    1    52 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.1           tcp dpt:587
    3   180 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.1           tcp dpt:993
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.1           tcp dpt:995
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.0.6           tcp dpt:19132
    0     0 ACCEPT     udp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.0.6           udp dpt:19132
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.2           tcp dpt:1194
    0     0 ACCEPT     udp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.1.2           udp dpt:1194
    0     0 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.10          tcp dpt:50455
    0     0 ACCEPT     udp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.10          udp dpt:50455
    0     0 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.7           tcp dpt:50456
    0     0 ACCEPT     udp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.7           udp dpt:50456
    4   188 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.0.4           tcp dpt:445
    0     0 ACCEPT     tcp  --  ens3   br-15b0c160fb5a  0.0.0.0/0            172.18.0.8           tcp dpt:3128
    0     0 ACCEPT     udp  --  ens3   br-2ae1c2fd8727  0.0.0.0/0            172.21.0.2           udp dpt:9987
    0     0 ACCEPT     tcp  --  ens3   br-2ae1c2fd8727  0.0.0.0/0            172.21.0.2           tcp dpt:10011
    0     0 ACCEPT     tcp  --  ens3   br-2ae1c2fd8727  0.0.0.0/0            172.21.0.2           tcp dpt:30033
    1    60 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.6           tcp dpt:80
    2    92 ACCEPT     tcp  --  ens3   br-182544ce66dd  0.0.0.0/0            172.20.0.6           tcp dpt:443
    0     0 ACCEPT     all  --  docker0 ens3    0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain DFWRS_INPUT (1 references)
 pkts bytes target     prot opt in     out     source               destination
   12   555 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            state INVALID
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED
 1724  103K ACCEPT     all  --  br-b08c9ac12933 *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-480a3d62c58f *       0.0.0.0/0            0.0.0.0/0
    1    60 ACCEPT     all  --  br-182544ce66dd *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-e4e931f0d1f6 *       0.0.0.0/0            0.0.0.0/0
   32  1920 ACCEPT     all  --  br-15b0c160fb5a *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-2ae1c2fd8727 *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  br-683fab2c3c96 *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  docker0 *       0.0.0.0/0            0.0.0.0/0
```
```
filter = [
    "-P FORWARD DROP",
	"-P INPUT DROP",
	"-P OUTPUT  ACCEPT",
	"-A INPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT",
	"-A INPUT -i lo -j ACCEPT",
	"-A OUTPUT -o lo -j ACCEPT",
	"-A INPUT -p icmp --icmp-type echo-request -j ACCEPT",
	"-A INPUT -p tcp --dport 22 -m conntrack --ctstate NEW,ESTABLISHED -j ACCEPT",
	"-A INPUT -p tcp --dport 25 ! -s 127.0.0.0/8 -j DROP",
	"-A INPUT -p tcp --dport 25 ! -s 111.11.11.11 -j DROP",
]
```

**Comment by pitkley** [comment]
Ahh, sorry, my fault:

1. The way dfw actually gets external traffic to the Docker containers is through the NATting features of iptables. If you execute `iptables -t nat -nvL`, you will see the `DFWRS_PREROUTING` chain that destination-NATs the traffic onto the internal Docker IPs. This chain is evaluated _before_ the `INPUT` chain, after which it will pass through the `FORWARD` chain instead, i.e. your rule never matches.

    What this effectively means is that you have to append the rule to the `FORWARD` chain, rather than the `INPUT` chain.

2. Since you are now inverting the range, you cannot have two source-rules `DROP`ping traffic to port 25. Your public IP `111.11.11.11` will never match `127.0.0.0/8`, and would fail. I think you'd have to comma-separate the two CIDRs.

This means that your entire rule should probably be:

```
-A FORWARD -p tcp --dport 25 ! -s 127.0.0.0/8,111.11.11.11/32 -j DROP
```

**Comment by cybermcm** [comment]
1) yes `-t nat` shows that PREROUTING comes before INPUT
2) not working,  if I use more than one IP dfw can't start, because of
>iptables-restore failed: ... not allowed with multiple source or destination IP addresses

**Comment by pitkley** [comment]
Argh, that's annoying 😕. I'll see if I can implement an alternative solution for your use-case, but I unfortunately have no direct suggestion you could apply to fix it in the meantime. I'll keep you up-to-date!

**Comment by cybermcm** [comment]
thank you for spending your time on this feature request (I think now it's a feature request ;-))

**Comment by pitkley** [comment]
Quick update: I have implemented a potential solution, and pushed it to Docker Hub under the tag `pitkley/dfw:gh89`. See [here](https://github.com/pitkley/dfw/blob/a592180f81c015ccbe9f0989fab82c7e5390d9b4/resources/test/conf-file.toml#L47-L52) on how you can configure the source IP-ranges for which you want to expose a port:

```toml
[[wider_world_to_container.rules]]
network = "network"
dst_container = "dst_container"
expose_port = 22
external_network_interface = "eni"
source_cidr = ["192.0.2.1/32", "192.0.2.2/32"]
```

If you build the image yourself, you can find the code in the [`112-allow-restricting-dnat-gh89` branch](https://github.com/pitkley/dfw/tree/112-allow-restricting-dnat-gh89).

Please let me know if that works correctly 👍.  If it does, I'll release v0.5.1 proper with this feature included!

**Comment by cybermcm** [comment]
Thank you for this quick fix, although it's currently not working
toml:
```
[[wider_world_to_container.rules]]
network = "compose"
dst_container = "mail"
expose_port = [587, 993, 995]

[[wider_world_to_container.rules]]
network = "compose"
dst_container = "mail"
expose_port = 25
external_network_interface = "ens3"
source_cidr = ["127.0.0.0/8", "111.11.11.11"]
```
I can see the IPs in iptables in the DFWRS_FORWARD chain -> perfect, but there is still on rule in place:
```
Chain DFWRS_PREROUTING (1 references)
 pkts bytes target     prot opt in     out     source               destination
    4   240 DNAT       tcp  --  ens3   *       0.0.0.0/0            0.0.0.0/0            tcp dpt:25 to:172.18.1.1:25
```
so port 25 is still open to the public

**Comment by pitkley** [comment]
The `FORWARD` rule is what decides if to forward or not in the end. Since your `FORWARD` default policy is `DROP`, the port should not be publicly reachable. Have you verified this using a port checker? Maybe also check it with [MXToolbox's SMTP checker](https://mxtoolbox.com/diagnostic.aspx), which might be more reliable to get an answer in this specific instance.

**Comment by pitkley** [comment]
Okay, found it. The default-policy of the `container_to_container` configuration-section sets the policy on the `FORWARD` chain -- which is obviously needed to be set to `ACCEPT` if you don't want to impede container-to-container communication.

I'll update my fix and will let you know once you can retest! 👍

**Comment by pitkley** [comment]
@cybermcm I have updated the image and branch to additionally enforce the source-IP in `PREROUTING`. Can you retest? 🙂

**Comment by cybermcm** [comment]
Just tested, works perfectly! The desired IPs have access, public IPs are blocked.
Thank you again for this quick fix and your time!

**Comment by pitkley** [comment]
Great, thank you so much for confirming. I have released dfw 0.5.1, which is being built by Docker Hub right now. I expect it to be available somewhere in the next 60-90 minutes, such that you can depend on it in the future (`latest` is obviously also an option!). :+1:
